text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,882 @@
+```rust
+// At call-site, there seem to be three comma-separated places,
+// while the function has five parameter slots
+takes_five(0, ...tup3, 4)
+    
+// Desugared, the call above looks like this:
+takes_five(0, tup3.0, tup3.1, tup3.2, 4);
+//            ^^^^^^^^^^^^^^^^^^^^^^^
+// The whole tuple of three fields was unpacked into the
+// underlined parameter slots.
+}
+```
+
+## Non-Trivial Cases and Interactions
+
+Basically, argument unpacking an expression of one of the allowed collection types desugars into the exhaustive consecutive field accesses in the order the elements would be accessed with `.idx`/`[idx]`, starting from `idx`' value `0`. The subchapters below discuss cases where this rule of thumb alone is insufficient.
+
+### Empty Collections
+
+When unpacking a unit tuple struct, the unit type, or an empty array, desugaring simply emits no field accesses. Essentially, the result is the same as if no arguments were passed at the point where a zero-length collection was unpacked.
+
+Since this is dead code, a lint is emitted.
+
+### Generic Parameters
+
+When function parameters are generic, using `<T>`, `impl` or `dyn`, exactly the same should happen as when the arguments are passed by hand. I.e., the argument's type must be *compatible* with the parameter's type. Just as when entering that argument manually with a field access expression.
+
+### References and Mutability in Function Parameters
+
+Rust's current syntax for calling functions that take one parameter by value, by reference, or by mutable reference requires the developer to be explicit about what they want:
+```rust
+fn ret_one_arg() -> u8 {
+    123
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn use_one_refmutarg(p: &mut u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(ret_one_arg());
+    use_one_refarg(&ret_one_arg()); // `&` needed to compile!
+    use_one_refmutarg(&mut ret_one_arg()); // `&mut` needed to compile!
+}
+```
+
+Rust also requires the developer to explicitly dereference references:
+```rust
+const CONST_NUMBER: u8 = 42;
+
+fn ret_one_refarg() -> &'static u8 {
+    &CONST_NUMBER
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(*ret_one_refarg()); // `*` needed to compile!
+    use_one_refarg(ret_one_refarg());
+}
+```
+
+Explicitly indicating varying degrees of (de)reference status or mutability on arguments being unpacked does not follow from the proposed syntax in any straightforward way. Thus, although it limits the usefulness of the feature, the design for such possibility is left out of the scope of this proposal. Consequently, the code will only compile if passing the arguments one by one with the corresponding field access expressions would compile.
```

Comment:
I agree that this should be left out of the MVP. But some discussion in the future possibilities section may be advisable.

Here is a possible syntax (which may have been proposed before):

```rust
let foo = (1, 2, 3);

func(&...foo);
// is equivalent to
func(&foo.0, &foo.1, &foo.2);
```